rust-runtime/aws-smithy-checksums/src/body.rs

```diff
@@ -0,0 +1,342 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use crate::http::{new_from_algorithm, HttpChecksum};
+
+use aws_smithy_http::body::SdkBody;
+use aws_smithy_http::header::append_merge_header_maps;
+
+use bytes::Bytes;
+use http::{HeaderMap, HeaderValue};
+use http_body::SizeHint;
+use pin_project_lite::pin_project;
+
+use std::fmt::Display;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+pin_project! {
+    /// A `ChecksumBody` will read and calculate a request body as it's being sent. Once the body has
+    /// been completely read, it'll append a trailer with the calculated checksum.
+    pub struct ChecksumBody<InnerBody> {
+            #[pin]
+            body: InnerBody,
+            checksum: Option<Box<dyn HttpChecksum>>,
+    }
+}
+
+impl ChecksumBody<SdkBody> {
+    /// Given an `SdkBody` and a `Box<dyn HttpChecksum>`, create a new `ChecksumBody<SdkBody>`.
+    pub fn new(body: SdkBody, checksum: Box<dyn HttpChecksum>) -> Self {
+        Self {
+            body,
+            checksum: Some(checksum),
+        }
+    }
+}
+
+impl http_body::Body for ChecksumBody<SdkBody> {
+    type Data = Bytes;
+    type Error = aws_smithy_http::body::Error;
+
+    fn poll_data(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Result<Self::Data, Self::Error>>> {
+        let this = self.project();
+        match this.checksum {
+            Some(checksum) => {
+                let poll_res = this.body.poll_data(cx);
+                match &poll_res {
+                    Poll::Ready(Some(Ok(data))) => checksum.update(&data),
+                    _ => (),
+                }
+
+                poll_res
+            }
+            _ => Poll::Ready(Some(Err(Box::new(Error::Taken)))),
+        }
+    }
+
+    fn poll_trailers(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<Option<HeaderMap>, Self::Error>> {
+        let this = self.project();
+        let poll_res = this.body.poll_trailers(cx);
+
+        if let Poll::Ready(Ok(maybe_inner_trailers)) = poll_res {
+            let checksum_headers = if let Some(checksum) = this.checksum.take() {
+                checksum.headers()
+            } else {
+                return Poll::Ready(Err(Box::new(Error::Taken)));
+            };
```

Comment:
let's return `Poll::Ready(Ok(None))` in this case—I think that's going to work with more downstream implementations

Reply:
we can also totally remove the error